lib/omniauth/strategies/openid_connect.rb

```diff
@@ -98,6 +102,23 @@ def config
         @config ||= ::OpenIDConnect::Discovery::Provider::Config.discover!(options.issuer)
       end
 
+      # @override
+      # Called before both of request_phase and callback_phase
+      def setup_phase
+        super
+
+        if configured_response_type != 'code' &&
+            configured_response_type != 'id_token token'
+          raise ArgumentError, 'Not supported response_type'
+        end
+
+        if configured_response_type == 'id_token token' &&
```

Comment:
`response_type` can also be:

1. `token`
2. `token id_token token`

Reply:
No. `token` (raw OAuth 2.0) MUST NOT be used for the authentication purpose. There are many explanations, for example, see: http://www.thread-safe.com/2012/01/problem-with-oauth-for-authentication.html

Acceptable:
 - `code` - Authorization Code Flow
 - `id_token token` - Implicit Flow
 - `code id_token` - Hybrid Flow -- Financial-grade API Security Profile 1.0 - Part 2: Advanced

However, I think i will implement the Implicit Flow first. After finishing, I will consider whether to implement the Hybrid Flow.

Reply:
This method name `setup_phase` is confusing; maybe `validate_response_type!` is better. My point is there are many types of `response_type` values, and it's not clear to me that this is checking specific types for different flows.

We may want to link: https://openid.net/specs/oauth-v2-multiple-response-types-1_0-03.html

Reply:
`setup_phase` is a override method. It's OK.
As you say, there are various combinations of `response_type`. However, they cannot be implemented without checking the use cases and risks one by one.
If you want to implement the Hybrid Flow, I won't stop it and welcome you to do it.

Reply:
I see your security note about `id_token`, but this gem supported that configuration in the past (e.g. https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-the-sign-in-request). I'm a bit hesitant to make this change now.

Reply:
There are too many things to point out and I am at a loss.
 - The Implicit Flow of this branch does not force the public key verification required by the specification and does not secure. So I'm trying to make it.
 - The compatibility with what does not work. What does that mean?
 - As stated in the link you shared, Azure AD disables the Implicit Flow by default.
 - I'm using Azure AD, so I checked. You can enable the implicit permissions on the Azure AD console. However, in the description there, `id_token token` is specified. In the case of the Hybrid Flow, enabling only `id_token` is instructed.
 - The Hybrid Flow is not the scope of this patch sequence.

Furthermore, if you still want to get `id_token` only, OK, I consider implementing it. Let me do to accept only `id_token` in the next patch part 2 and implement the public key validation in patch part 3.

Reply:
@stanhu 
Are there any additions? For reference, I would link to the implementation guide.
https://openid.net/specs/openid-connect-implicit-1_0.html